### ISSUE TITLE
Unexpected semicolon added with `--no-semi`

### ISSUE THREAD

**Issue by mmiller42** [prettier/prettier#2841]
**Prettier 1.7.0**
[Playground link](https://prettier.io/playground/#%7B%22content%22%3A%22const%20config%20%3D%20()%20%3D%3E%20%7B%5Cn%20%20%5B1%2C%202%2C%203%5D.forEach(n%20%3D%3E%20console.log(n))%5Cn%7D%22%2C%22options%22%3A%7B%22ast%22%3Afalse%2C%22bracketSpacing%22%3Atrue%2C%22doc%22%3Afalse%2C%22jsxBracketSameLine%22%3Afalse%2C%22output2%22%3Afalse%2C%22parser%22%3A%22babylon%22%2C%22printWidth%22%3A80%2C%22semi%22%3Afalse%2C%22singleQuote%22%3Afalse%2C%22tabWidth%22%3A2%2C%22trailingComma%22%3A%22none%22%2C%22useTabs%22%3Afalse%7D%7D)
```sh
--no-semi
```

**Input:**
```js
const config = () => {
  [1, 2, 3].forEach(n => console.log(n))
}
```

**Output:**
```js
const config = () => {
  ;[1, 2, 3].forEach(n => console.log(n))
}

```

**Expected behavior:**

There's an unexpected semicolon placed before the array definition. This is the expected result:

```diff
  const config = () => {
-   ;[1, 2, 3].forEach(n => console.log(n))
+   [1, 2, 3].forEach(n => console.log(n))
  }

```

Interestingly, this particular semicolon is not rendered when `--no-semi` is removed.

**Comment by hawkrives** [comment]
You'll be interested in https://github.com/prettier/prettier/issues/1907#issuecomment-305913157, https://github.com/prettier/prettier/issues/736#issuecomment-291934981, and https://github.com/prettier/prettier/issues/736#issuecomment-291930139.

TL;DR: Prettier's `--no-semi` flag errs on the side of caution and code editability, inserting semicolons before `(`, `[`, and `backtick` characters to allow people to move the lines around easily without having to worry about ASI triggering because you moved the line.
